### ISSUE TITLE
Don't use `AtomicBool` in `seg_queue::Segment`

### ISSUE THREAD

**Issue by cybergeek94** [aturon/crossbeam#19]
The `ready` field of `seg_queue::Segment` is *horribly* wasteful:

        ready: [AtomicBool; SEG_SIZE],

`AtomicBool` uses `usize` internally, so on 64-bit platforms and `SEG_SIZE = 32`, that's 256 bytes, 2048 bits, for only 32 bits' worth of information. *In each segment*.

I think we can do *much* better here, using `AtomicUsize` as the storage for an atomically accessed bit-set. Since we only need 32 bits of information (true or false for each cell), we can get away with a single `AtomicUsize`, a whopping 248 bytes' reduction in size *per segment* for little additional cost elsewhere. For very long queues, this could mean a significant savings in memory usage, which also means more cache-friendliness.

There might very well be some nuances I'm overlooking, but it seems as though the semantics should remain almost exactly the same. The one thing I can see is increased contention across threads, but I don't know if that'd be an issue or not. Clearly some benchmarks are needed here.

**Comment by cybergeek94** [comment]
I did some basic benchmarks, and using a single `AtomicUsize` is about twice as slow as an array of `AtomicBool`. I expected it to be faster due to better cache locality.

**Comment by gnzlbg** [comment]
Do you know why? (Or could you post the benchmarks?)

I would guess that you are hitting false sharing. Basically for atomic
variables, cache lines must have a consistent state between the caches of
each CPU core.

That is, writes to the cache line containing a single `AtomicUsize` must be
synchronized between core caches. If two writes to it happen in different
caches at the same "synchronization step", the operation is retried until
this succeeds. This can make microbenchmarks much slower. Furthermore, any
other write to the cache line where the `AtomicUsize` lives might be
responsible for this, so it would be useful to know if there is anything on
that cache line that is also hot. Adding some padding (a cache line width)
before an after the `AtomicUsize` should be enough to study this.

If you are really hitting false sharing, the `AtomicBool` array might be
hitting it too. It would be interesting to study if having a single
`AtomicBool` per cache line can further improve the performance at the cost
of ~4-8x more memory.

TL;DR: unless you are modifying a group of atomic variables from the same
CPU core (i.e. your threads are pin), you generally want your atomic
variables to be allocated in different cache lines if they are under a lot
of contention from multiple threads (which is typically the case in
microbenchmarks, and might not be representative of real applications).

**Comment by cybergeek94** [comment]
@gnzlbg Here is the gist, this is the `lib.rs` of a Cargo project with `lazy_static = "*"` in its dev dependencies:

https://gist.github.com/cybergeek94/6ea304b19781b670a26e

I just now added the padding to `AtomicBitSet`, but it didn't make a difference at all. 8 `usize`s should be 64 bytes, the length of a x86 cache line, on my machine, so that's taking a whole cache line for itself.

These are the bench results:

    test test::bench_atomic_bit_set    ... bench:       1,186 ns/iter (+/- 171)
    test test::bench_atomic_bool_array ... bench:         485 ns/iter (+/- 8)

It's curious. I'm not an assembly guru but I don't think I see any atomic operations actually taking place in `bench_atomic_bool_array`. `bench_atomic_bit_set` contains `lock` instructions which, according to a Google search. lock the cache line like you said. The array doesn't do that.

**Comment by cybergeek94** [comment]
It just occurred to me that I'm not even testing concurrent access. Maybe
I'll add some duplicate benches for some contention.

**Comment by gnzlbg** [comment]
>@gnzlbg Here is the gist, this is the lib.rs of a Cargo project with lazy_static = "*" in its dev dependencies:
>
>https://gist.github.com/cybergeek94/6ea304b19781b670a26e

But that's all serial right? Then the problem cannot be about false sharing. 

Can you post the assembly? The bit set solution cannot be vectorized, but the array solution can and probably is being vectorized.

> I just now added the padding to AtomicBitSet, but it didn't make a difference at all. 8 usizes should be 64 bytes, the length of a x86 cache line, on my machine, so that's taking a whole cache line for itself.

Yes, in x86 a cache line is 8 * 64bits. The size of `AtomicBitSet` without padding is only 64bit, so in general it will share a cache line with other stuff. The test of making sure that `AtomicBitSet` is "alone" in a single cache line is only to discard that the slow down comes from anything else it might be sharing the cache line with. 

I forgot to mention before that padding is not enough. The `AtomicBitSet` can still be allocated between two cache lines. So besides padding, one needs to require `AtomicBitSet` to be aligned to a 64byte boundary (its memory address must be a multiple of the cache line size). 

If this makes no difference at all, then the only thing we know is that _iff_ this is due to false sharing, the false sharing is occurring only within the `AtomicBitSet` and nothing else is involved in it.

**Comment by schets** [comment]
Just looking at the bitset implementation, there's going to be really bad CAS contention on the atomic or -  Even a successful CAS can be reasonably expensive, and failed ones are outright bad.

You can run [this](https://github.com/schets/contention_test) for some benchmarks on your machine of how various x86 atomics fare under different levels on contention - long story short, its not good with no contention and really bad with contention.

IMHO, each element in the array should be a struct containing the ready variable and the element, that will get some better cache separation and reads on one will usually read the other as well.

**Comment by cybergeek94** [comment]
Yeah, I realize that my intentions were misguided. Forget I said anything.

**Comment by schets** [comment]
Actually, I think they are right - that array is really wasteful and there's likely a better way to do this.
